semver/src/lib.sk

```diff
@@ -0,0 +1,267 @@
+module Semver;
+
+class Version(
+  major: Int,
+  minor: Int,
+  patch: Int,
+  pre: Prerelease = Prerelease::kEmpty,
+  build: BuildMetadata = BuildMetadata::kEmpty,
+) uses Orderable {
+  static fun fromString(str: String): Version {
+    pos = Lexer.LexingPosition::create(str);
+
+    res = parse_version(pos);
+
+    if (!pos.atEnd()) {
+      invariant_violation("Unexpected character after version")
+    };
+
+    res
+  }
+
+  fun toString(): String {
+    res = `${this.major}.${this.minor}.${this.patch}`;
+    if (!this.pre.isEmpty()) {
+      !res = `${res}-${this.pre}`
+    };
+    if (!this.build.isEmpty()) {
+      !res = `${res}+${this.build}`
+    };
+
+    res
+  }
+
+  // Cf. https://semver.org/:
+  // > Build metadata MUST be ignored when determining version precedence.
+  fun compare_precedence(other: Version): Order {
+    (this.major, this.minor, this.patch, this.pre).compare(
+      (other.major, other.minor, other.patch, other.pre),
+    )
+  }
+}
+
+class VersionReq(comparators: Array<Comparator>) uses Show {
```

Comment:
Is there a spec for this?

Reply:
It is (heavily) inspired from [cargo](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html), but no official spec that I know of.